go/dbagent/dbagent.go

```diff
@@ -0,0 +1,360 @@
+	} else {
+		query = fmt.Sprintf("GRANT %s ON %s TO '%s'@'%s';", grant, object, username, host)
+	}
+	_, err := ExecuteQuery(query)
+	return err
+}
+
+func ManageReplicationUser() error {
+	var err error
+	version, err := GetMySQLVersion()
+	if err != nil {
+		log.Errore(err)
+	}
+	if !UserExists(config.Config.MySQLReplicationUser) {
+		if version == "5.7" {
+			err = CreateUser(config.Config.MySQLReplicationUser, "%", config.Config.MySQLReplicationPassword)
+			if err != nil {
+				return err
+			}
+		} else {
+			err = GrantUser(config.Config.MySQLReplicationUser, "%", "REPLICATION SLAVE", "*.*", config.Config.MySQLReplicationPassword)
+		}
+	}
+	if !HasGrant(config.Config.MySQLReplicationUser, "Repl_slave_priv") {
+		err = GrantUser(config.Config.MySQLReplicationUser, "%", "REPLICATION SLAVE", "*.*", "")
+	}
+	query := fmt.Sprintf(`FLUSH PRIVILEGES;`)
+	_, err = ExecuteQuery(query)
+	if err != nil {
+		return log.Errore(err)
+	}
+	return err
+}
+
+func GenerateBackupForUsers(users []string) (backup string, err error) {
+	if db, err := OpenConnection(); err == nil {
+		for _, user := range users {
+			// small hack cos MySQL 5.6 doesn't have DROP USER IF EXISTS. USAGE is synonym for "no privileges"
+			backup += fmt.Sprintf("GRANT USAGE ON *.* TO %s IDENTIFIED BY 'temporary_password'; DROP USER %s;", user, user)
+			query := fmt.Sprintf("SHOW CREATE USER %s", user)
+			res, err := sqlutils.QueryResultData(db, query)
+			if err != nil {
+				log.Errore(err)
+			} else {
+				backup += res[0][0].String + ";\n"
+			}
+			query = fmt.Sprintf("SHOW GRANTS FOR %s", user)
+			res, err = sqlutils.QueryResultData(db, query)
+			if err != nil {
+				log.Errore(err)
+			} else {
+				for r := range res {
+					backup += res[r][0].String + ";\n"
+				}
+			}
+		}
+	}
+	return backup, err
+}
+
+func StartSlave(sourceHost string, sourcePort int, logFile string, position string, gtidPurged string) error {
+	var err error
+	if len(logFile) > 0 && len(position) > 0 {
+		query := fmt.Sprintf(`CHANGE MASTER TO MASTER_LOG_FILE='%s', MASTER_LOG_POS=%s;`, logFile, position)
+		_, err = ExecuteQuery(query)
+		if err != nil {
+			return log.Errore(err)
+		}
+	}
+	if len(gtidPurged) > 0 {
+		// as we are restoring new slave and if we have GTIDs we may have @@GLOBAL.GTID_EXECUTED not empty and we won't be able to set @@GLOBAL.GTID_PURGED
+		// so we need to do RESET MASTER on slave
+		query := `RESET MASTER;`
+		if _, err = ExecuteQuery(query); err != nil {
+			return log.Errore(err)
+		}
+		query = fmt.Sprintf(`SET @@GLOBAL.GTID_PURGED='%s';`, gtidPurged)
+		_, err = ExecuteQuery(query)
+		if err != nil {
+			return log.Errore(err)
+		}
+		query = `CHANGE MASTER TO MASTER_AUTO_POSITION = 1;`
+		_, err = ExecuteQuery(query)
+		if err != nil {
+			return log.Errore(err)
+		}
+	}
+	query := fmt.Sprintf(`CHANGE MASTER TO MASTER_HOST='%s', MASTER_USER='%s', MASTER_PASSWORD='%s', MASTER_PORT=%d, MASTER_CONNECT_RETRY=10;`,
```

Comment:
Why `10`? I'm curious. I use `86400` or `1000000`.

Reply:
Removed in babfa38 as duplicate of orchestrator functionality